Ruby 3.4 and 4.0 compatibility (require Ruby 2.3)

sylvain-8422 opened the pull request:
**Avoid Ruby 3.4 deprecation warnings and eventual Ruby 4.0 failures by**:

- **requiring Ruby 2.3** ;
- **and marking a few `""` literals as being mutable (using `+""`)**.

Examples of warnings:

```
marc-1.3.0/lib/marc/xml_parsers.rb:97: warning: literal string will be frozen in the future (run with --debug-frozen-string-literal for more information)
marc-1.3.0/lib/marc/xml_parsers.rb:98: warning: literal string will be frozen in the future (run with --debug-frozen-string-literal for more information)
marc-1.3.0/lib/marc/xml_parsers.rb:99: warning: literal string will be frozen in the future (run with --debug-frozen-string-literal for more information)
```

I only added `+` to most `""`s without much thought, more changes may be necessary in order to be fully compatible with Ruby 3.4 and up.

Proposed changes should be safe as long as at least Ruby 2.3 is used.

sylvain-8422 committed:
Ruby 3.4 and 4.0 compatibility (require Ruby 2.3)

sylvain-8422 commented:
If requiring at least Ruby 2.3 is an issue, every `+""` could be replaced by `"".dup` instead, then it would work with 2.2. 😬

(Ruby 2.3 was released almost 10 years ago. 🧓)

jrochkind commented:
Ah there is actually a fix here! Thank you @sylvain-8422 !

I don't use ruby-marc myself, so I don't know if I'm the right person to approve this -- but I do use things that have ruby-marc as a dependency, so have an interest!

Is there anyone else feeling like a maintainer on this right now?

billdueber commented:
I'm happy to merge this as-is, unless anyone else disagrees.

jrochkind commented:
Looks like we don't have any automatic CI in ruby-marc?

sylvain-8422 commented:
The _perfect solution_ would be to make sure that the whole _gem_ works fine with `# frozen_string_literal: true` headers.

But, in the meantime, if requiring Ruby 2.3 is not an issue, the merge of this _pull request_ would be a good start, as all the changes should be harmless.

dmoles commented:
@billdueber LGTM, unless we want to also add 3.5 to the [test matrix](https://github.com/ruby-marc/ruby-marc/blob/main/.github/workflows/ruby.yml) as part of this (assuming GitHub supports it)?

@jrochkind we're supposed to have CI, I'm not sure why it's not triggering for every PR. I've never looked (closely) at the GH action myself, though.

dmoles commented:
@sylvain-8422 personally I'm fine with dropping 2.2 support; it's only been unsupported for 7½ years, and anyone who still really needs it probably isn't actively developing and could just pin to an earlier ruby-marc version.

dmolesUC commented:
(oops, posted those comments from the wrong account, but yes, it's me)

jrochkind commented:
i thought the workflow might be disabled, but it does not seem to be. I don't know why CI isn't running. I guess ideally for such a wide PR we really would have CI.  Don't know if anyone has bandwidth to try to figure out what's going on. 

Dropping ancient rubies seems fine.

sylvain-8422 commented:
**As for automatic CI, the GitHub workflow probably did not run for my pull request because it is from a remote branch.**

You need to enable a configuration so that it runs on branches from forks (disabled by default for security reasons).

**Look in the repository settings, under "actions".**

jrochkind commented:
If you know what's going on and how to do that, want to throw it into this PR? Or is it somethign that has to be done in repo settings web UI, not in config file? We could prob make you a committer with access to settings too if you need it.

sylvain-8422 commented:
It might not be the problem, but it would be the first thing to look at : **from the repository Web UI -> Settings -> Actions**.

<img width="661" height="119" alt="image" src="https://github.com/user-attachments/assets/a91fc3da-d2f2-4428-9cf2-1f8d05d67c69" />

and maybe

<img width="542" height="54" alt="image" src="https://github.com/user-attachments/assets/842552d3-6375-4c3d-bb8e-f63801489a23" />

rsinger commented:
Unfortunately, that doesn't seem to be available, at least not in that wording. Maybe somebody else already did it. 

From what I can tell, there _should_ be an option for one of us to manually run the actions, per https://docs.github.com/en/actions/how-tos/manage-workflow-runs/approve-runs-from-forks but that's not showing up for me. If somebody had changed the repo setting after the last branch update, @sylvain-8422 you might need to push a change to trigger the event. 

I've created #112 which I _think_ would let us run the CI pipeline manually on this repo for any branch, which might be a solution if we can't get that UI option to appear on this PR.

billdueber commented:
Hey folks -- getting at this after a long while. I'm gonna fix the test matrix/gh actions if I can, deal with the tests manually if need be, and get this merged and pushed out as v1.4. Objections?

billdueber committed:
Pulling down PR#110 to try to get it to run in GH test matrix

billdueber committed:
Updates to deal with CI and frozen string warnings;

  - Update CI to work on `workflow_dispatch` and add ruby 4.0 to the test matrix.
  - Remove flag `--without-documentation` fromn `bundle install` call in github CI
  - Remove call to JSON.fast_generate to avoid warnings
  - Only force marc8 string to binary encoding if it's not already
